hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java

```diff
@@ -306,6 +311,58 @@ public StorageStatistics provide() {
 
   }
 
+  private TransferManager createUploadTransferManager(Configuration conf) {
+    int maxThreads = conf.getInt(UPLOAD_MAX_THREADS, DEFAULT_UPLOAD_MAX_THREADS);
+    if (maxThreads < 2) {
+      LOG.warn(UPLOAD_MAX_THREADS + " must be at least 2: forcing to 2.");
+      maxThreads = 2;
+    }
+    long keepAliveTime = longOption(conf, KEEPALIVE_TIME,
+            DEFAULT_KEEPALIVE_TIME, 0);
+
+    uploadUnboundedThreadPool = new ThreadPoolExecutor(
+            maxThreads, Integer.MAX_VALUE,
+            keepAliveTime, TimeUnit.SECONDS,
+            new LinkedBlockingQueue<Runnable>(),
+            BlockingThreadPoolExecutorService.newDaemonThreadFactory(
+                    "s3a-upload-unbounded"));
+
+    TransferManagerConfiguration transferConfiguration =
+            new TransferManagerConfiguration();
+    transferConfiguration.setMinimumUploadPartSize(partSize);
+    transferConfiguration.setMultipartUploadThreshold(multiPartThreshold);
+
+    TransferManager uploads = new TransferManager(s3, uploadUnboundedThreadPool);
+    uploads.setConfiguration(transferConfiguration);
+    return uploads;
+  }
+
+  private TransferManager createCopyTransferManager(Configuration conf) {
```

Comment:
this is almost the same as the previous one apart from the conf fields and names; should be a common method

Reply:
Re-factored. Most of this was moved into `LazyTransferManager`. I abstracted the determination of # of threads into a common method.